### ISSUE TITLE
OS3/m68k issue: build fails with "Internal compiler error: program cc1 got fatal signal 11"

### ISSUE THREAD

**Issue by jens-maus** [jens-maus/amissl#4]
When trying to compile our latest openssl-1.1.x branch sources compilation for m68k-amigaos fails with a fatal compiler error:

```
[...]
m68k-amigaos-gcc -DOPENSSL_NO_DYNAMIC_ENGINE -DOPENSSLDIR="\"AmiSSL:\"" -DENGINESDIR="\"AmiSSL:engines\"" -m68020-60 -msoft-float -DNO_INLINE_VARARGS -D__NO_NET_API -DB_ENDIAN -DOPENSSL_SYS_AMIGA -D__amigaos3__ -DOPENSSL_NO_STDIO -I../../include -I../../include/netinclude -W -Wall -O1 -g -gstabs -DBN_DEBUG -DCONF_DEBUG -DDEBUG  -resident32 -DAMISSL_COMPILE -I../../libcmt/include  -Iinclude -I../../openssl -I../../openssl/crypto/include -I../../openssl/include -Icrypto/include -c -o crypto/dh/dh_rfc5114.o ../../openssl/crypto/dh/dh_rfc5114.c
In file included from ../../openssl/crypto/dh/dh_rfc5114.c:63:
../../openssl/crypto/include/internal/bn_dh.h:6: warning: `extern' is not at beginning of declaration
m68k-amigaos-gcc: Internal compiler error: program cc1 got fatal signal 11
Makefile:3402: recipe for target 'crypto/dh/dh_rfc5114.o' failed
```

Looking at the output of `dmesg` cc1 seems to crash with a segmentation fault:

```
[1299419.719277] cc1[16431]: segfault at 14 ip 0000000008198a8f sp 00000000ffed7538 error 4 in cc1[8048000+1e1000]
```

After further investigation I was able to manually try to investigate this issue with debugging in `gdb`:

(1) Compile `dh_rfc5441.c` sources with `-savetemps` so that the compiler saves the *.i and *.s file accordingly:
```
$ cd build_os3/openssl
$ m68k-amigaos-gcc -DOPENSSL_NO_DYNAMIC_ENGINE -DOPENSSLDIR="\"AmiSSL:\"" -DENGINESDIR="\"AmiSSL:engines\"" -m68020-60 -msoft-float -DNO_INLINE_VARARGS -D__NO_NET_API -DB_ENDIAN -DOPENSSL_SYS_AMIGA -D__amigaos3__ -DOPENSSL_NO_STDIO -I../../include -I../../include/netinclude -W -Wall -O1 -g -gstabs -DBN_DEBUG -DCONF_DEBUG -DDEBUG  -resident32 -DAMISSL_COMPILE -I../../libcmt/include  -Iinclude -I../../openssl -I../../openssl/crypto/include -I../../openssl/include -Icrypto/include -c -o crypto/dh/dh_rfc5114.o ../../openssl/crypto/dh/dh_rfc5114.c -save-temps
```

(2) Run `cc1` manually to trigger the issue:

```
$ /opt/m68k-amigaos/lib-clib2/gcc-lib/m68k-amigaos/2.95.3/cc1 dh_rfc5114.i -dumpbase dh_rfc5114.c -da -m68020-60 -msoft-float -g -gstabs -O1 -W -Wall -version -fbaserel32 -o /tmp/ccNJom9s.s
[...]
Segmentation fault (core dumped)
```

(3) Run the same command in a `gdb` session:

```
$ gdb /opt/m68k-amigaos/lib-clib2/gcc-lib/m68k-amigaos/2.95.3/cc1
[...]
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from /opt/m68k-amigaos/lib-clib2/gcc-lib/m68k-amigaos/2.95.3/cc1...done.
(gdb)
```

(4) Now run `cc1` with the `run` gdb command to actually trigger the issue:

```
(gdb) run dh_rfc5114.i -dumpbase dh_rfc5114.c -da -m68020-60 -msoft-float -g -gstabs -O1 -W -Wall -version -fbaserel32 -o /tmp/ccNJom9s.s
[...]
Program received signal SIGSEGV, Segmentation fault.
0x08198a8f in amigaos_put_in_text (decl=0x84fae54)
    at /home/damato/projekte/Amiga/amigaos-cross-toolchain/.build-m68k/sources/gcc-2.95.3/gcc/config/m68k/amigaos.c:63
63	  return (TREE_INT_CST_HIGH (TYPE_SIZE (type)) == 0
(gdb)
```

(5) run `where` to see the backtrace:

```
(gdb) where
#0  0x08198a8f in amigaos_put_in_text (decl=0x84fae54)
    at /home/damato/projekte/Amiga/amigaos-cross-toolchain/.build-m68k/sources/gcc-2.95.3/gcc/config/m68k/amigaos.c:63
#1  0x080b15e8 in make_decl_rtl (decl=0x84fae54, asmspec=0x0, top_level=1)
    at /home/damato/projekte/Amiga/amigaos-cross-toolchain/.build-m68k/sources/gcc-2.95.3/gcc/varasm.c:783
#2  0x0804beaf in rest_of_decl_compilation (decl=0x84fae54, asmspec=0x0, top_level=1, at_end=0)
    at /home/damato/projekte/Amiga/amigaos-cross-toolchain/.build-m68k/sources/gcc-2.95.3/gcc/toplev.c:3526
#3  0x081ac431 in finish_decl (decl=0x84fae54, init=0x0, asmspec_tree=0x0)
    at /home/damato/projekte/Amiga/amigaos-cross-toolchain/.build-m68k/sources/gcc-2.95.3/gcc/c-decl.c:4103
#4  0x081a04ac in yyparse () at c-parse.y:1015
#5  0x0804b761 in compile_file (name=0xffffd5bd "dh_rfc5114.i")
    at /home/damato/projekte/Amiga/amigaos-cross-toolchain/.build-m68k/sources/gcc-2.95.3/gcc/toplev.c:3270
#6  0x0804eeb1 in main (argc=16, argv=0xffffd404)
    at /home/damato/projekte/Amiga/amigaos-cross-toolchain/.build-m68k/sources/gcc-2.95.3/gcc/toplev.c:5454
```

Obviously there seems to be a problem in code generation that needs to be investigated in the gcc-295.3 sources.

**Comment by jens-maus** [comment]
@Futaura please note my latest checkin to the openssl-1.1.x sources. This seems to workaround the GCC crashes we were seeing. Would be great to identify the root cause and fix it in GCC 2.95.3, though.

**Comment by jens-maus** [comment]
The latest version of the used cross compiler toolchain (cahirwpz/amigaos-cross-toolchain) doesn't require our workaround anymore. Thus this ticket can be closed after having reverted our workaround.
